libraries/postgres/src/Database.js

```diff
@@ -721,6 +744,49 @@ class Database {
 
     return Buffer.concat([b1, b2]);
   }
+
+  /**
+   * Given a row and an array of serialization functions, generate a signature
+   * container using the last serialization in the array.
+   */
+  sign({row, signingSerializations}) {
+    const ser = signingSerializations.length - 1;
+    const v = 0;
+    const {id: kid} = this.signingKeyring.currentKey('hmac-sha512');
+    const sig = this.calculateSignature({row, kid, v, ser, signingSerializations});
+    return {v, ser, kid, sig};
+  }
+
+  /**
+   * Given a row, signature, and serializations, verify the signature.  This
+   * performs a fixed-time comparison to avoid timing attacks on the signature.
+   * Returns true if the signature is valid.
+   */
+  verifySignature({row, signature, signingSerializations}) {
+    const sig = this.calculateSignature({
+      row,
+      kid: signature.kid,
+      v: signature.v,
+      ser: signature.ser,
+      signingSerializations,
+    });
+    return fixedTimeComparison(Buffer.from(signature.sig, 'base64'), Buffer.from(sig, 'base64'));
+  }
+
+  /**
+   * Helper for sign and verifySignature
+   */
+  calculateSignature({row, kid, v, ser, signingSerializations}) {
```

Comment:
Similar to https://github.com/taskcluster/taskcluster/pull/3138/files#r449564731 why not use [Sign](https://nodejs.org/api/crypto.html#crypto_class_sign) and [Verify](https://nodejs.org/api/crypto.html#crypto_class_verify) from the Node crypto lib here?

Reply:
If that produces matching output, that's totally fine -- but it looks like it uses asymmetric-key encryption, where the existing signing is based on HMAC, so symmetric.  We need to maintain compatibility with existing signatures.  We could migrate to asymmetric-key later using the `algo` parameter to in the signature container, but since we'd always carry the public and private keys around together, I'm not sure there's much advantage (and there's cost to deploying two much larger keys).

@ajvb I think the "our own implementation" characterization is referring to the fixed-time comparison, yes?  The use of HMAC here seems pretty standard and non-clever.